sound/soc/bcm/es9023_dac.c

```diff
@@ -0,0 +1,259 @@
+	    sample_freq < 192000 && sample_freq % 8000 == 0) {
+		if (sample_bits == 16 || sample_bits == 24)
+			bclk_ratio = 50;
+		else if (sample_bits == 32)
+			bclk_ratio = 100;
+	}
+
+	dev_dbg(rtd->dev, "%s: frequency=%u, format=%s, sample_bits=%u, "
+		"physical_bits=%u, channels=%u. Setting bclk_ratio=%u.\n",
+		__func__, sample_freq, snd_pcm_format_name(format),
+		sample_bits, physical_bits, channels, bclk_ratio);
+
+	return snd_soc_dai_set_bclk_ratio(cpu_dai, bclk_ratio);
+}
+
+/* machine stream operations */
+static struct snd_soc_ops snd_rpi_es9023_dac_ops = {
+	.hw_params = snd_rpi_es9023_dac_hw_params,
+};
+
+static char *snd_rpi_es9023_dac_get_card_name(struct device_node *np)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(es9023_dac_variants); i++) {
+		if (of_device_is_compatible(np,
+					    es9023_dac_variants[i].compatible))
+			return es9023_dac_variants[i].card_name;
+	}
+
+	return es9023_dac_variants[0].card_name;
+}
+
+static char *snd_rpi_es9023_dac_get_dai_name(struct device_node *np)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(es9023_dac_variants); i++) {
+		if (of_device_is_compatible(np,
+					    es9023_dac_variants[i].compatible))
+			return es9023_dac_variants[i].dai_name;
+	}
+
+	return es9023_dac_variants[0].dai_name;
+}
+
+static char *snd_rpi_es9023_dac_get_dai_stream_name(struct device_node *np)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(es9023_dac_variants); i++) {
+		if (of_device_is_compatible(np,
+					    es9023_dac_variants[i].compatible))
+			return es9023_dac_variants[i].dai_stream_name;
+	}
+
+	return es9023_dac_variants[0].dai_stream_name;
+}
+
+static int snd_rpi_es9023_dac_probe(struct platform_device *pdev)
+{
+	struct device_node *codec_np, *i2s_np, *np = pdev->dev.of_node;
+	struct es9023_dac_priv *priv;
+	int ret = 0;
+
+	i2s_np = of_parse_phandle(np, "i2s-controller", 0);
+	if (!i2s_np) {
+		dev_err(&pdev->dev, "%s: failed to find i2s-controller\n",
+			__func__);
+		ret = -EINVAL;
+		goto i2s_end;
+	}
+
+	codec_np = of_parse_phandle(np, "audio-codec", 0);
+	if (!codec_np) {
+		dev_err(&pdev->dev, "%s: failed to find audio-codec\n",
+			__func__);
+		ret = -EINVAL;
+		goto end;
+	}
+
+	priv = devm_kzalloc(&pdev->dev, sizeof(*priv), GFP_KERNEL);
+	if (!priv) {
+		ret = -ENOMEM;
+		goto end;
+	}
+
+	/* priv->dai.cpu_dai_name = "bcm2708-i2s.0"; */
+	priv->dai.cpu_of_node = i2s_np;
+
+	/* priv->dai.platform_name = "bcm2708-i2s.0"; */
+	priv->dai.platform_of_node = i2s_np;
+
+	/* priv->dai.codec_name = "es9023-codec"; */
+	/* priv->dai.codec_dai_name = "es9023-hifi"; */
+	priv->dai.codec_of_node = codec_np;
+
+	priv->dai.dai_fmt = SND_SOC_DAIFMT_I2S |
+			    SND_SOC_DAIFMT_NB_NF |
+			    SND_SOC_DAIFMT_CBS_CFS;
+
+	priv->dai.ops = &snd_rpi_es9023_dac_ops;
+
+	priv->card.dai_link = &priv->dai;
+	priv->card.dev = &pdev->dev;
+	priv->card.num_links = 1;
+	priv->card.owner = THIS_MODULE;
+
+	priv->bclk_ratio_int_div = of_property_read_bool(np,
+					"es9023-dac,bclk-ratio-int-div");
+
```

Comment:
Wouldn't it be smaller, simpler and more efficient to have one function that determines the board variant, returning a `const struct es9023_dac_variant*` from which you can pick the relevant field rather than four virtually identical functions to perform the lookup and then return one value from the structure?

Or just read the strings from DT.